src/coreclr/src/jit/hwintrinsicxarch.cpp

```diff
@@ -528,6 +528,118 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_Vector128_AsVector:
+        {
+            assert(sig->numArgs == 1);
+
+            if (getSIMDVectorRegisterByteLength() == YMM_REGSIZE_BYTES)
+            {
+                // Vector<T> is 32-bytes, so we should treat this as a call to Vector128.ToVector256
+                return impBaseIntrinsic(NI_Vector128_ToVector256, clsHnd, method, sig, mustExpand);
+            }
+
+            assert(getSIMDVectorRegisterByteLength() == XMM_REGSIZE_BYTES);
+
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD.
+
+            retNode = impSIMDPopStack(retType);
+            SetOpLclRelatedToSIMDIntrinsic(retNode);
+            assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+            break;
+        }
+
+        case NI_Vector128_AsVector2:
+        case NI_Vector128_AsVector3:
+        case NI_Vector128_AsVector4:
+        {
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD or the
+            // return type is a smaller TYP_SIMD that shares the same register.
+
+            retNode = impSIMDPopStack(retType);
+            SetOpLclRelatedToSIMDIntrinsic(retNode);
+            assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+            break;
+        }
+
+        case NI_Vector128_AsVector128:
```

Comment:
`AsVector128` is a bit more complex.

For `Vector4` to `Vector128` it should be simple as it is `TYP_SIMD16` to `TYP_SIMD16`

The other two (which are currently unhandled) I'd like some input on how to handle.
If they are in register, it can be a nop. Likewise, if they were spilled to the stack, they can be a nop.
However, if it is a field or an array element, we have to read 8 or 12-bytes exactly (I probably need to propose `Unsafe` versions so we don't have to `zero-init` the upper element as well).

Should we carry the `NI_Vector128_AsVector128` through to codegen and decide what to do there (nop vs actual read, depending on location) or is there something clever we can do earlier in the compilation?

Reply:
CC. @CarolEidt for input here

Reply:
It would be nice to be able to handle this in a "generic" way - that is, using a `CAST` from one type to another that does the appropriate widening.  However, I don't think the current handling of casts is "generic" in a good way. So, I'd probably vote for carrying the conversion through to `Lowering` and codegen.

Reply:
> that is, using a CAST from one type to another that does the appropriate widening.

Such a cast shouldn't be needed, much like we don't need a cast from `short` to `int` after a `short` indir. A SIMD12 indir really produces a SIMD16 value.

Reply:
I'm going to leave this to a future PR for now. `Vector2/3` to `Vector128<float>` is going to be left as a managed implementation which explicitly zeroes the unused upper elements.

In particular I want to get some more testing and samples done for this and don't want to block the other changes.